el-patch-template

haji-ali opened the pull request:
This is an implementation of the ideas in #50. I put it as a pull request to discuss the implementation and how it can be improved/refined.

All the suggested examples by @raxod502 should work. Currently, the only limitation is that `...` does not work as expected inside `el-patch-concat`. In particular, I would expect it to match a substring so that one does not have to write the full string if one wants to replace a single word in a long string. The implementation of such a functionality is not difficult but I wanted to get some feedback before complicating the code further.

Also, currently `el-patch-template` simply returns the patch form rather than yanking it to the kill-ring or executing it. For now, this can be done with

```
(kill-new (format "%S" (el-patch-template ...))
(eval (el-patch-template ...))
```
respectively. (In this instance `...` should be filled. `el-patch-template` is not magic :) ) 

Any feedback on the syntax or my lisp is much appreciated (I am not an expert with lisp).

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,433 @@
+              (cadr qkeyword-name)))
+     (setq patched (el-patch-template--apply definition
+                                             ',templates))
+     ;; NOTE: Unfortunately `el-patch-deftype-alist' doesn't save the
+     ;; macro name so we have to assume that it is `el-patch-*'
```
I wonder if `el-patch-deftype-alist' should save the macro-name if it is provided to `el-patch-deftype`.

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,433 @@
+                                        table literal in-string))
+          (throw 'no-match nil))))))
+
+(defun el-patch-template--match-p (form template)
+  "Matches TEMPLATE to FORM. TEMPLATE may contain `...' which
+greedily match any number of form. Match is succesful if a
+partial list of FORM, starting from the beginning matches
+TEMPLATE. The return value is the number of forms in FORM which
+match template or nil if a match is not possible."
```
This is a fast version of `el-patch-template--process` that does not expect `el-patch-*` forms but does handle `...`. It is used as a fast matching function, so it must be called with a resolved template. When a match is successful, the function `el-patch-template--process` is then called with the original template (which includes `el-patch-*`).

haji-ali commented:
String substitution now works as well. Now an `el-patch-template` with a string can be short an sweet

```elisp
(el-patch-template
 (defun (el-patch-swap restart-emacs radian-new-emacs))
 (el-patch-concat
   (el-patch-swap
     "Restart Emacs."
     "Start a new Emacs session without killing the current one.")
   ...
   (el-patch-swap "restarted" "started")
   ...
   (el-patch-swap "restarted" "started")
   ...
   (el-patch-swap "restarted" "started")
   ...)
 (el-patch-swap
   (save-buffers-kill-emacs)
   (restart-emacs--launch-other-emacs restart-args)))
```

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+
+;; Author: Al Haji-Ali <abdo.haji.ali@gmail.com>
+;; Created: 1 March 2021
+;; Homepage: http://www.randomoid.com
```
This should probably be

```suggestion
;; Homepage: https://github.com/raxod502/el-patch
```

since I think it is for the package not hte author.

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+;; Keywords: extensions
+;; Package-Requires: ((emacs "25"))
+;; SPDX-License-Identifier: MIT
+;; Version: 0.0.1
```
We'll want to make this part of the same package, I'm guessing:

```suggestion
;; Version: 2.3.1
```

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+  `(el-patch-template--impl (quote ,keyword-name)
+                            (quote ,templates)))
+
+(provide 'el-patch-template)
```
```suggestion
(provide 'el-patch-template)

;; Local Variables:
;; indent-tabs-mode: nil
;; End:

;;; el-patch-template.el ends here
```

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+
+(defun el-patch-template--process-el-patch
+    (form template &optional match next-step-fn table)
+  "Processes an el-patch statement. Arguments are the same as
+`el-patch-template--match'. Assume thats template is a cons whose
```
The style guide for Elisp docstrings says to put argument names in all uppercase, and to write in the imperative verb tense ("Do X" rather than "Does X").

```suggestion
  "Process an el-patch statement. Arguments are the same as
`el-patch-template--match'. Assumes that TEMPLATE is a cons whose
```

Similar comments apply to the other docstrings, I think.

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+                  (cddr patch)))))
+
+(defmacro el-patch-template (keyword-name &rest templates)
+  "KEYWORD-NAME is cons whose car is a type which can be any type
+from `el-patch-deftype-alist' and the cdr is the name of the form
```
Is this true? I think it's a list of two elements, rather than a cons. (The latter would be `(defun . name-of-function)`.)

haji-ali replied:
Right, corrected.

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+from `el-patch-deftype-alist' and the cdr is the name of the form
+to be patched. Looks for all forms which match a template in
+TEMPLATES and processes them by matching all `...' forms against
+the souce code of the form, while keeping the `el-patch-*'
```
```suggestion
the source code of the form, while keeping the `el-patch-*'
```

raxod502 commented on el-patch-template.el:
```diff
@@ -0,0 +1,513 @@
+                    (unless (plist-get ptemplate :matched)
+                      (error
+                       "At least one template did not match any form"))))))
+    ;; NOTE: Unfortunately `el-patch-deftype-alist' doesn't save the
+    ;; macro name so we have to assume that it is `el-patch-*'
```
Is that true? I believe it does save the `macro-name` argument; but it won't be in the alist if it wasn't provided. In other words, you can try looking it up, and if it's not there, then that means to use the default of `el-patch-%S`.

haji-ali replied:
Ah, I misread your code. Yes, that's true.

raxod502 reviewed (commented):
I made a few small comments, although I didn't look through the whole implementation. If it works, that's the most important thing. I'm happy to advise on more specific parts if there are places you think should be improved.

Side note---we should add `el-patch-template.el` to the `for_compile` and `for_checkdoc` lists in Makefile.

haji-ali commented on Makefile:
```diff
@@ -62,7 +62,7 @@ checkindent: ## Ensure that indentation is correct
 	    emacs -Q --batch \
 	        --eval "(setq inhibit-message t)" \
 	        --eval "(load (expand-file-name \"el-patch.el\") nil t)" \
```
Should I load `el-patch-template.el` here too? I am not sure why `el-patch.el` itself is loaded.

raxod502 replied:
The reason for `el-patch.el` being loaded is to get the macros defined. This is because macros are indented differently than regular functions, and if the macro definition isn't loaded, then Emacs does not know how to indent it correctly. You only need to load `el-patch-template.el` if that file defines a macro which has indentation rules different than a regular function.

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,554 @@
+(defun el-patch-template--impl (keyword-name templates)
+  "Apply the templates to the definition of an elisp construct.
+
+The actual implementation of `el-patch-template', accepts the
+same arguments but quoted."
```
Lint complains about the documentation here (missing argument names). Should I copy the documentation from `el-patch-template`? Seems redundant.

raxod502 replied:
Checkdoc is a bit silly sometimes. What I've done in this case is something like this:

```elisp
(defun el-patch-template--impl (keyword-name templates)
  "Apply the templates to the definition of an elisp construct.
The meaning of KEYWORD-NAME and TEMPLATES is the same as in
`el-patch-template' (which see), but here they need to be quoted
since they are passed as regular function arguments."
```

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,554 @@
+                 1))))))
+
+(defun el-patch-template--any-p (definition ptemplates &optional up-to)
+  "Return t if any template in PTEMPLATES matches any form in DEFINITION.
```
`checkdoc` complains about "matches" here, but it refers to the argument rather than the function!

raxod502 replied:
Yeah, this is silly. You can fix it by adding

```elisp
;; checkdoc-verb-check-experimental-flag: nil
```

to the `Local Variables` list at the bottom of the file.

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,554 @@
+     (t
+      (or (when-let ((symbol (gethash template table))
+                     (symbol-next-step
+                      (lambda (new-match remainder-form)
+                        ;; First put the match in the table.
+                        (let ((old-entry (gethash template table)))
+                          (puthash template
+                                   (cons
+                                    (car symbol)
+                                    new-match)
+                                   table)
+                          ;; Then process the next step, adding the
+                          ;; template to the match
+                          (condition-case _
+                              (funcall next-step-fn
+                                       (if match
+                                           (cons match template)
+                                         template)
+                                       remainder-form)
+                            ('no-match
+                             ;; Ultimately, the matching did not
+                             ;; work, so undo the symbol resolution
+                             (puthash template old-entry table)
+                             ;; and rethrow
+                             (throw 'no-match nil)))))))
```
The processing functions `el-patch-template--process-el-patch` and `el-patch-template--process`  accept a `next-step-fn` function handle which is called with the matched and unmatched forms. This is important for `el-patch-*` directives because at this point the matched forms are processed to be a valid `el-patch-*` directive. This works well, but the functions has these lambda definitions which are sometimes long and might make the function hard to read. 

Do you think these definitions are fine? Or should I write functions like `el-patch-template--process-next-step/let` which return these lambda definitions given some parameters? This would also remove unnecessary capturing of parameters, if any.

raxod502 replied:
Yeah, I think this is perfectly good. Of course we can always refactor later if it turns out to be necessary, but I think it looks great as is! I wouldn't worry about lexical capturing---that should have negligible impact, even if it feels unclean.

haji-ali commented:
Thanks @raxod502. I've tested different forms and everything seems to work. Hopefully I didn't miss a use-case.

I've tried to follow the documentation standards as you suggested. I added a few questions in the code review that I would appreciate your input in.

Also, regarding your suggestions in #50. How about I create two versions:
- `el-patch-template` evaluates the `el-patch` definition. If in init file prints a warning about the evaluation speed. If in compile time expands into the `el-patch-*` definition (Not 100% sure how to do that last one and would appreciate pointers).
- `el-patch-template*` which simply returns the `el-patch` definition unless called interactively in which case yanks the definition in the kill-ring.

I don't know if/how I can combine all of these into one version depending on context.

raxod502 commented:
> How about I create two versions

Yes, two entry points seems right. I gave some thought to how to make the behavior of these functions the least surprising to the user, and came up with this proposal:

* We have a macro called `el-patch-define-template` which doesn't actually compute anything. Rather it simply puts its arguments into a shared variable, just like defining a regular `el-patch` does (but this variable would be separate than the one used to keep track of all existing `el-patch` forms).
* We have an interactive command called `el-patch-insert-template` which selects a template from the ones defined in the shared variable, like `el-patch-ediff-patch`, computes its expansion, and then inserts the expansion into the current buffer at point.
* We can also have a macro called `el-patch-define-compiletime-template`, which will take the same arguments as `el-patch-define-template`, and do the same thing (namely, add arguments to a shared variable for later retrieval). However, it will also have an additional effect, which is to check whether macroexpansion is happening during compilation or not. If it is, then it will compute the template expansion and then macroexpand to the expanded template. If not, then it will skip that step and print a warning.

Then our two use cases look like this:

* *Simple, standard usage:* Put an `el-patch-template` form into your init-file. Then call `el-patch-insert-template` to generate and insert the expanded patch into your init-file each time it's updated.
* *Advanced usage with byte-compilation:* Put an `el-patch-define-compiletime-template` form into your init-file, and ensure that you byte-compile your init-file before loading it.

> Also not sure how to do that last one

Here ya go!

```elisp
(defmacro magic-macro ()
  (message "This message gets printed during macroexpansion")
  `(progn
     (message "This message gets printed at runtime")
     ,(if (bound-and-true-p byte-compile-current-file)
          '(message "The macro was expanded during byte-compilation")
        '(message "The macro was expanded at runtime"))))

(magic-macro)
```

And then you can see the results as follows:

```
% emacs -Q --batch -l magic.el
This message gets printed during macroexpansion
This message gets printed at runtime
The macro was expanded at runtime

% emacs -Q --batch -f batch-byte-compile magic.el
This message gets printed during macroexpansion

% emacs -Q --batch -l magic.elc                  
This message gets printed at runtime
The macro was expanded during byte-compilation
```

haji-ali commented:
OK, I implemented those idea. I like the interface you suggested, but I find the fact that `el-patch-define-compiletime-template` throws an error during runtime a bit counterintuitive.

Instead, I made it so that `el-patch-define-compiletime-template` defines and evaluates a template when called in runtime. A warning is emitted in that case if `el-patch-warn-on-eval-template` is non-nil. If `el-patch-define-compiletime-template` is called in compile-time, the template is defined and its resolution is macro-expanded.

In addition to  `el-patch-insert-template`, I also added a function `el-patch-eval-template` which evaluates a template.

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,682 @@
+              ;; Add template definition to the macro expansion so
+              ;; that when the compiled file is executed the template
+              ;; definition is still accessible.
+              `(when (featurep 'el-patch-template)
+                 ;; Only define the template if el-patch-template is
+                 ;; loaded
+                 (el-patch--define-template (quote ,type-name)
+                                            (quote ,templates)))
```
I figured we would want to have the template defined if `el-patch-template` is loaded so that the user can interact with the template afterwards. Otherwise, only the template resolution is executed.

I could also resolve the template and execute that expansion instead of the el-patch definition if `el-patch` is not loaded. Although this does increase the size of the compiled file (both the patch and its resolution must be included in the macro expansion). Maybe there's a better way? I couldn't figure out how el-patch does this.

haji-ali commented on el-patch-template.el:
```diff
@@ -0,0 +1,688 @@
+        `(progn
+           ;; In case el-patch-template is not loaded then simply
+           ;; set `el-patch--templates' to an empty hash
+           (setq el-patch--templates
+                 (or (bound-and-true-p el-patch--templates)
+                     (make-hash-table :test 'equal)))
+           ;; Add template definition to `el-patch--templates'
+           (puthash (quote ,(car type-name))
+                    (cons (quote ,(cadr type-name))
+                          (quote ,templates))
+                    (or (gethash (quote ,resolved-name) el-patch--templates)
+                        (puthash (quote ,resolved-name)
+                                 (make-hash-table :test #'equal)
+                                 el-patch--templates)))
```
I figured we would want to have the template definition saved whether `el-patch-template` is loaded or not so that the user can interact with the template afterwards.

I don't know if I also need to handle the case when `el-patch` is not loaded or if that is handled by `el-patch` itself.

raxod502 commented:
> Sorry for the notifications! I just learned about drafts in github and converted the request to a draft

Oh, no worries, I disable notifications for newly pushed commits in any case :)

> added a function `el-patch-eval-template`

Makes sense!

> To my mind, this is the most intuitive behavior but the function name might need rethinking

Yeah, that seems pretty reasonable. Maybe could rename to `el-patch-define-and-eval-template` or something.

> I figured we would want to have the template definition saved

Yep yep, that makes sense.

> I don't know if I also need to handle the case when `el-patch` is not loaded

Yeah, I think we do need to handle that. It should be fairly straightforward though, I think you just need to autoload any variables that need to be defined for the macroexpansion to execute successfully even if `el-patch` isn't loaded, e.g.:

https://github.com/raxod502/el-patch/blob/5e823dc9a29e3be22597d93912f06119c38030d6/el-patch.el#L137-L141

haji-ali commented on el-patch.el:
```diff
@@ -640,12 +640,14 @@ is the Lisp form, read from the buffer at point."
+                   (save-excursion
+                     (goto-char defun-point)
+                     (read (current-buffer)))))
+       (when defun-buffer
+         (kill-buffer defun-buffer)))))
```
I added this logic to kill the buffer of the source file. I think keeping the buffer open (for templates or patches) is unnecessary.

raxod502 replied:
Yeah, that seems good to me. Since we have the hack just above to ensure `el-patch` creates a new buffer rather than reusing an existing buffer, we know that this will not accidentally kill one of the user's currently open buffers, so it should be safe.

haji-ali commented:
I've added validation functions for templates (stolen from el-patch) and greatly simplified the implementation to avoid exceeding `max-lisp-eval-depth` when the template is very long.

In my tests, having my templates in my init file did not slow down my startup time significantly (especially when I autoload them). However, I am still debugging an issue where getting the source from `.el.gz` files fails sometimes; it seems that `jka` is somehow not fully loaded when my init file is executed. This might be related to my Radian config being loaded in `early-init.el` (see https://github.com/raxod502/radian/issues/474), but I am not sure yet.

> Yeah, I think we do need to handle that. It should be fairly straightforward though, I think you just need to autoload any variables that need to be defined for the macroexpansion to execute successfully even if `el-patch` isn't loaded

Hmm... Maybe I am misunderstanding something here. Here's what I am thinking a use-case might be:

- User generates a compiled file `use-compiled-file.elc` which contains el-patch templates and/or patches.
- User puts `(require 'use-compiled-file)` in their init file. This loads their patches, i.e., their function definitions are evaluated. Loading this compiled module does not require el-patch or el-patch-template being loaded, nor does it load them. In fact, `el-patch` and `el-patch-template` need not be even available.
- User later loads `el-patch` and `el-patch-template` (interactively or otherwise) and then is able to execute `el-patch-*` functions to validate their patches and/or templates.

In `el-patch-template`, the way I did this is by adding the template definition to `el-patch--template`, making sure to define it (without loading `el-patch-template`) if it wasn't bound. I thought a similar solution would need to be implemented for `el-patch`, i.e., the patch definition is added to `el-patch--patches` (without loading `el-patch`) and its resolution is macroexpanded.

Is this what you also have in mind? If so, maybe I am fundamentally misunderstanding autoloading. I thought autoloading variables would cause the package containing them to be loaded whenever they are used. Is that not the case?

raxod502 commented:
> having my templates in my init file did not slow down my startup time significantly

OK, good to know. I guess most people don't care about having literally the smallest number of milliseconds of startup time, and it's fine to support having the templates expanded during startup, as long as there's also an option to not do that.

> It turns out that Radian sets file-name-handler-alist to nil

Oops, my bad :)

> Here's what I am thinking a use-case might be

Your use case is the exact same one I have in mind.

> In el-patch-template, the way I did this is by adding the template definition to el-patch--templates, making sure to define it (without loading el-patch-template) if it wasn't bound

That's a fine approach as well. It should have the same practical effect as what I suggested (autoloading the variable definition `el-patch--templates`).

> rely on the user generating the autoload file and loading it correctly

By "the autoload file" do you mean the generated file which contains the autoloads of the `el-patch` package itself? If so, then I think it should be safe to assume that will be loaded correctly. `package.el`, `straight.el`, and other package managers should all take care of this automatically.

That said, it is true that if the user is installing and loading `el-patch` manually, then your solution would degrade more gracefully than mine. That would be a perfectly reasonable reason to go with your way, if you prefer it.

Note that the definition of `el-patch--templates` in the autoload file doesn't actually contain any user data: it's literally just `(defvar el-patch--templates nil)` if the variable is autoloaded. It's still up to the macroexpanded code to push values onto that list at runtime.

> I couldn't find where the resolution of the el-patch is macroexpanded.

So, suppose you have this in your init-file:

```elisp
(el-patch-defun company-statistics--load ()
  "Restore statistics."
  (load company-statistics-file 'noerror
        (el-patch-swap nil 'nomessage)
        'nosuffix))
```

Then it macroexpands to this (courtesy of https://github.com/joddie/macrostep):

```elisp
(el-patch--definition
 (defun company-statistics--load nil "Restore statistics."
        (load company-statistics-file 'noerror
              (el-patch-swap nil 'nomessage)
              'nosuffix)))
```

And then the `el-patch--definition` macro expands again to:

```elisp
(progn
  (puthash 'defun
           '(defun company-statistics--load nil "Restore statistics."
                   (load company-statistics-file 'noerror
                         (el-patch-swap nil 'nomessage)
                         'nosuffix))
           (or
            (gethash 'company-statistics--load el-patch--patches)
            (puthash 'company-statistics--load
                     (make-hash-table :test
                                      (function equal))
                     el-patch--patches)))
  (el-patch--stealthy-eval
   (defun company-statistics--load nil "Restore statistics."
          (load company-statistics-file 'noerror 'nomessage 'nosuffix))
   "This function was patched by `el-patch'."))
```

Where as you can see there is a `puthash` that installs the full text of the patch into the variable `el-patch--patches`. This would normally fail if `el-patch.el` has not yet been loaded, since the variable would be unbound (not even `nil`). However, since we autoload `el-patch--patches`

https://github.com/raxod502/el-patch/blob/5e823dc9a29e3be22597d93912f06119c38030d6/el-patch.el#L137-L141

then the package manager will copy the following into the autoload file:

```elisp
(defvar el-patch--patches (make-hash-table :test 'equal) "\
Hash table of patches that have been defined.
The keys are symbols naming the objects that have been patched.
The values are hash tables mapping definition types (symbols
`defun', `defmacro', etc.) to patch definitions, which are lists
beginning with `defun', `defmacro', etc.

Note that the symbols are from the versions of patches that have
been resolved in favor of the modified version, when a patch
renames a symbol.")
```

which should be loaded before anything else happens in user configuration, even if `el-patch.el` is not actually loaded. Then, by time we get to evaluating the `el-patch-defun` form (or its macroexpanded/compiled version), the variable is defined and can be updated with the patch text.

Does that answer your question?

haji-ali commented:
> Does that answer your question?

Thanks for taking the time to write that answer. Yes, everything is clear now. 
I didn't realize that macros keep expanding. It that case, the implementation should be complete now.

I added a section in the README and I think the branch is ready to be merged. Let me know if you think other changes are needed.

raxod502 commented:
Thank you so much for your great work and patience! :rose:
